### PR TITLE
[327] Server should allow "Restart in Debug"

### LINKED ISSUE #327: Server should allow "Restart in Debug"
steps:
1. ASSERT: have a server
1. ASSERT: have it "(Started)"
1. EXEC: try to debug the server

Result:
You cannot have it run in debug from here. There's no "Restart in Debug", you first have to "Stop" and then you get an entry "Debug" where you can start the server in debug mode.

### PR DESCRIPTION
**Pull request by lstocchi**
it fixes #327 

This solution adds a new button in the context menu. 

Signed-off-by: Luca Stocchi lstocchi@redhat.com

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -81,6 +82,11 @@
         "title": "Restart",
         "category": "Servers"
       },
+      {
+        "command": "server.restartDebug",
+        "title": "Restart in Debug",
```
Comment (adietish) [inline package.json]: somehow "Restart Debugging" would look nicer to me.

**Comment by adietish** [comment]
@lstocchi another question is if we want to implement a different approach in terms of actions available in the context menu: 
image
In this PR we add another entry to the context menu and it'll most likely keep on growing. In this light we should maybe have a "Restart" that simply restarts the current mode (if in debug, restart into debug, if in run, restart into run)

**Comment by lstocchi** [comment]
> @lstocchi another question is if we want to implement a different approach in terms of actions available in the context menu:
> image
> In this PR we add another entry to the context menu and it'll most likely keep on growing. In this light we should maybe have a "Restart" that simply restarts the current mode (if in debug, restart into debug, if in run, restart into run)

Yes this was also my concern but i am not sure which could be the best solution. I proposed this one as first because it is the same thing i found on eclipse. 
Wouldn't it be a limitation to allow the user to only restart the server in the current mode? I have no idea honestly.

**Comment by adietish** [comment]
> > @lstocchi another question is if we want to implement a different approach in terms of actions available in the context menu:
> > image
> > In this PR we add another entry to the context menu and it'll most likely keep on growing. In this light we should maybe have a "Restart" that simply restarts the current mode (if in debug, restart into debug, if in run, restart into run)
> 
> Yes this was also my concern but i am not sure which could be the best solution. I proposed this one as first because it is the same thing i found on eclipse.
> Wouldn't it be a limitation to allow the user to only restart the server in the current mode? I have no idea honestly.

Sure, if we dont offer "Restart Debugging" we'd not allow users to switch from "Run" to "Debugging" in 1 step, they would have to first "Stop" and "Debug". I agree on this. On the other hand I fear with @robstryker that the number of items in the ctx menu gets too large to stay clear. The idea therefore is to sacrifice this 1-step offering to a more concise ctx-menu. I'm not completely sure if that's the right choice but worth considering seriously.
@robstryker @jeffmaury thoughts?

**Comment by jeffmaury** [comment]
I agree that we should take care about the number of items in the context menu. However, I think this can be done in a separate issue as the current state is acceptable.
However, I have one question: is the menu available if the server is already in debug mode ? Is this something we want to support ?

**Comment by adietish** [comment]
> I agree that we should take care about the number of items in the context menu. However, I think this can be done in a separate issue as the current state is acceptable.
> However, I have one question: is the menu available if the server is already in debug mode ? Is this something we want to support ?

It is, I would argue that this fulfills the use case is running in "Debug" and you want to restart it, into "Debug" again. We actually have the same with "Restart" when in "Run" mode.

**Comment by robstryker** [comment]
I think we should rename both actions as follows:
"Restart in run mode"
"Restart in debug mode"

The problems here are if you are in debug mode originally, you might assume restart simply means restart in debug mode. It's not clear it would change the mode to run.

**Comment by adietish** [comment]
> I think we should rename both actions as follows:
> "Restart in run mode"
> "Restart in debug mode"
> 
> The problems here are if you are in debug mode originally, you might assume restart simply means restart in debug mode. It's not clear it would change the mode to run.

Agree with mentioning the mode in both cases which adds clarity. But then I'd prefer shorter, more concise versions like (compare UX good practices which also suggest to use title capitalization: https://uxmovement.com/buttons/5-rules-for-choosing-the-right-words-on-button-labels/)

* "Restart Debugging"
* "Restart Running"

even though I admit that "Running" is not obvious. So maybe just drop the "in"
* "Restarting Debug Mode"
* "Restart Run Mode"

**Comment by adietish** [comment]
Once the labelling is corrected we should merge this, I believe. I filed the error when restarting into debug mode to a new issue: #346

**Comment by lstocchi** [comment]
I vote for 

- "Restart in Run Mode"
- "Restart in Debug Mode"

It's clearer for me. I'm not a native english speaker so maybe I miss something but it doesn't seem correct to use these two options below. @robstryker can be more helpful than me for sure.

- "Restarting Debug Mode" -> in this case we are not going to short the label anyway
- "Restart Run Mode"

**Comment by adietish** [comment]
@lstocchi we also need to fix the tests. I'm on it.

**Comment by adietish** [comment]
@lstocchi I fixed the tests: https://github.com/lstocchi/vscode-adapters/pull/3 
Please merge --rebase my PR before changing the labels.

**Comment by adietish** [comment]
testPR

**Comment by adietish** [comment]
testPR

**Comment by adietish** [comment]
weird, tests are fine locally but travis fails to build it.

File: test/debugInfo.test.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
+/*-----------------------------------------------------------------------------------------------
+ *  Copyright (c) Red Hat, Inc. All rights reserved.
+ *  Licensed under the EPL v2.0 License. See LICENSE file in the project root for license information.
+ *-----------------------------------------------------------------------------------------------*/
+import * as chai from 'chai';
+import * as chaipromise from 'chai-as-promised';
+import { ClientStubs } from './clientstubs';
+import { DebugInfo } from '../src/debugInfo';
+import { DebugInfoProvider } from '../src/debugInfoProvider';
```
Comment (jrichter1) [inline test/debugInfo.test.ts]: well, these files start with a capital D

**Comment by jrichter1** [comment]
> weird, tests are fine locally but travis fails to build it.

Delete the out folder and try again

File: test/debugInfo.test.ts (hunk 2 above)
Comment (adietish) [reply test/debugInfo.test.ts]: matches travis (build on mac succeeded while the one on linux failed). On Mac this doesn't matter while on Linux it does. (was caused by squashing, rebasing etc.)

**Pull request merged**